lib/Dialect/FIRRTL/Transforms/InferWidths.cpp

```diff
@@ -1236,6 +1236,52 @@ void InferenceMapping::constrainTypes(Value larger, Value smaller) {
   constrain(type, larger, smaller);
 }
 
+/// Establishes constraints to ensure the sizes in the `larger` type are greater
+/// than or equal to the sizes in the `smaller` type. The types do not have to
+/// be identical, but they must be similar in the sense that corresponding
+/// fields must have the same kind (scalars, bundles, vectors).
+///
+/// This function is used to apply partial connects.
+void InferenceMapping::partiallyConstrainTypes(Value larger, Value smaller) {
+  // Recurse to every leaf element and set larger >= smaller.
+  std::function<void(FIRRTLType, Value, unsigned, FIRRTLType, Value, unsigned)>
+      constrain = [&](FIRRTLType aType, Value a, unsigned aID, FIRRTLType bType,
+                      Value b, unsigned bID) {
+        if (auto aFlipType = aType.dyn_cast<FlipType>()) {
+          auto bFlipType = bType.cast<FlipType>();
+          // Flip the LHS and RHS.
+          constrain(bFlipType.getElementType(), b, bID,
+                    aFlipType.getElementType(), a, aID);
+        } else if (auto aBundle = aType.dyn_cast<BundleType>()) {
+          auto bBundle = bType.cast<BundleType>();
+          for (unsigned aIndex = 0, e = aBundle.getNumElements(); aIndex < e;
+               ++aIndex) {
+            auto aField = aBundle.getElements()[aIndex].name.getValue();
+            auto bIndex = bBundle.getElementIndex(aField);
+            if (!bIndex)
+              continue;
+            auto &aElt = aBundle.getElements()[aIndex];
+            auto &bElt = bBundle.getElements()[*bIndex];
+            constrain(aElt.type, a, aID + aBundle.getFieldID(aIndex), bElt.type,
+                      b, bID + bBundle.getFieldID(*bIndex));
+          }
+        } else if (auto aVecType = aType.dyn_cast<FVectorType>()) {
+          auto bVecType = bType.cast<FVectorType>();
+          constrain(aVecType.getElementType(), a, aID + 1,
+                    bVecType.getElementType(), b, bID + 1);
+        } else if (aType.isGround()) {
+          // Leaf element, look up their expressions, and create the constraint.
+          constrainTypes(getExpr(FieldRef(a, aID)), getExpr(FieldRef(b, bID)));
+        } else {
+          llvm_unreachable("Unknown type inside a bundle!");
+        }
+      };
+
+  auto largerType = larger.getType().cast<FIRRTLType>().stripFlip().first;
+  auto smallerType = smaller.getType().cast<FIRRTLType>().stripFlip().first;
+  constrain(largerType, larger, 0, smallerType, smaller, 0);
+}
```

Comment:
Can you add some tests that exercise the `flip` code path through this?

This looks correct to me, but I'd like to lock in the behavior. I ran this through the following and all the widths were inferred correctly. 😄

```
circuit Foo:
  module Foo:
    wire a0: {b: {c: UInt}}
    wire b0: {b: {c: UInt<1>}}

    a0 is invalid
    b0 is invalid

    a0 <= b0

    wire a1: {b: {flip c: UInt}}
    wire b1: {b: {flip c: UInt<1>}}

    a1 is invalid
    b1 is invalid

    b1 <= a1

    wire a2: {flip b: {c: UInt}}
    wire b2: {flip b: {c: UInt<1>}}

    a2 is invalid
    b2 is invalid

    b2 <= a2

    wire a3: {flip b: {flip c: UInt}}
    wire b3: {flip b: {flip c: UInt<1>}}

    a3 is invalid
    b3 is invalid

    a3 <= b3

  module Bar:
    wire a0: {b: {c: UInt}}
    wire b0: {b: {c: UInt<1>}}

    a0 is invalid
    b0 is invalid

    a0.b <= b0.b

    wire a1: {b: {flip c: UInt}}
    wire b1: {b: {flip c: UInt<1>}}

    a1 is invalid
    b1 is invalid

    b1.b <= a1.b

    wire a2: {flip b: {c: UInt}}
    wire b2: {flip b: {c: UInt<1>}}

    a2 is invalid
    b2 is invalid

    a2.b <= b2.b

    wire a3: {flip b: {flip c: UInt}}
    wire b3: {flip b: {flip c: UInt<1>}}

    a3 is invalid
    b3 is invalid

    b3.b <= a3.b

  module Baz:
    wire a0: {b: {c: UInt}}
    wire b0: {b: {c: UInt<1>}}

    a0 is invalid
    b0 is invalid

    a0.b.c <= b0.b.c

    wire a1: {b: {flip c: UInt}}
    wire b1: {b: {flip c: UInt<1>}}

    a1 is invalid
    b1 is invalid

    a1.b.c <= b1.b.c

    wire a2: {flip b: {c: UInt}}
    wire b2: {flip b: {c: UInt<1>}}

    a2 is invalid
    b2 is invalid

    a2.b.c <= b2.b.c

    wire a3: {flip b: {flip c: UInt}}
    wire b3: {flip b: {flip c: UInt<1>}}

    a3 is invalid
    b3 is invalid

    a3.b.c <= b3.b.c
```

Reply:
Yeah that's a good point. Thanks for the test snippet; let me add this to the test suite.